esrally/mechanic/supplier.py

```diff
@@ -694,6 +756,80 @@ def run(self, command, override_src_dir=None):
             raise BuildError(msg)
 
 
+class DockerBuilder:
+    def __init__(self, src_dir, build_jdk=None, log_dir=None, client=None):
+        self.client = client
+        self.src_dir = src_dir
+        self.build_jdk = build_jdk
+        self.log_dir = log_dir
+        self.logger = logging.getLogger(__name__)
+
+        try:
+            self.logger.info(self.client.version())
+        except docker.errors.APIError as e:
+            raise exceptions.SystemSetupError(
+                f"Docker installation not found: [{e}] To install or build from source, please install Docker."
+            )
+
+    def resolve_jdk_build_container_image(self, build_jdk: int) -> str:
+        """
+        Given a JDK major release version, find a suitable docker image to use for compiling ES & ES plugins from sources
+        :param build_jdk: The build JDK major version required to build Elasticsearch and plugins.
+        :return: The Docker image name and respective tag that satisfies the 'build_jdk' requirement.
+        """
+        logger = logging.getLogger(__name__)
+        # Temurin only maintains images for JDK 11, 17, 18 (at time of writing)
+        # older builds of ES may require specifics like JDK 14 etc, in which case we rely on the OpenJDK images
+        if build_jdk >= 17:
+            docker_image_name = f"eclipse-temurin:{build_jdk}-jdk"
+            logger.info("Build JDK version [%s] is >= 17, using Docker image [%s].", build_jdk, docker_image_name)
+        else:
+            docker_image_name = f"openjdk:{build_jdk}-jdk"
+            logger.info("Build JDK version [%s] is < 17, using Docker image [%s].", build_jdk, docker_image_name)
+        return docker_image_name
+
+    def build(self, commands):
+        build_command = "; ".join(commands)
+        self.run(build_command)
+
+    def run(self, command):
+        container_name = "esrally-source-builder"
+        io.ensure_dir(self.log_dir)
+        log_file = os.path.join(self.log_dir, "build.log")
+        try:
+            existing_container = self.client.containers.get(container_name)
+            existing_container.stop()
+            existing_container.remove()
+        except docker.errors.NotFound:
+            pass
+        # run the container in detatched mode
+        container = self.client.containers.run(
+            detach=True,
+            name=container_name,
+            image=self.resolve_jdk_build_container_image(self.build_jdk),
+            command=f'/bin/bash -c "for i in 1 2 3 4 5; do apt update -y; apt install git -y && break || sleep 15; done; '
```

Comment:
You can replace `1 2 3 4 5` with `{1..5}` to make it less ugly.

Reply:
While at it you could add a comment explaining what's the reason for the useradd's